# Séance projet 1

## 1. Préparation du travail
**Question 1 :** Importer ci-dessous tous les modules que vous jugerez nécessaires.

In [12]:
# Imports nécessaires

import pandas as pd

import csv

# Nous n'aurons pas besoin d'importer d'autre librairies pour cette séance

**Question 2 :** Charger le jeu de données qui se trouve dans le fichier `JO.csv` et afficher les 5 premières lignes pour vérifier que le chargement a été effectué correctement. 

Pensez à télécharger au préalable le fichier `JO.csv` et à le placer dans le même dossier que votre notebook.

Si vous avez importé le notebook sur Google Colab, il faut exécuter la cellule ci-dessous au préalable pour télécharger votre jeu de données dans votre environnement de travail sur Colab (inutile dans le cas où vous exécutez le notebook sur votre propre machine) :

In [2]:
!wget https://git.unistra.fr/dbernhard/isdd_data/raw/master/JO.csv

dyld: Library not loaded: /usr/local/opt/gettext/lib/libintl.8.dylib
  Referenced from: /usr/local/bin/wget
  Reason: image not found


In [3]:
# Chargement du jeu de données et affichage des 5 premières lignes

raw_data = pd.read_csv('JO.csv')

raw_data.head(n=5)

ID           Name Sex   Age  Height  Weight   Team  NOC        Games  \
0     1      A Dijiang   M  24.0   180.0    80.0  China  CHN  1992 Summer   
1     2       A Lamusi   M  23.0   170.0    60.0  China  CHN  2012 Summer   
2   602  Abudoureheman   M  22.0   182.0    75.0  China  CHN  2000 Summer   
3  1463      Ai Linuer   M  25.0   160.0    62.0  China  CHN  2004 Summer   
4  1464      Ai Yanhan   F  14.0   168.0    54.0  China  CHN  2016 Summer   

   Year  Season            City       Sport  \
0  1992  Summer       Barcelona  Basketball   
1  2012  Summer          London        Judo   
2  2000  Summer          Sydney      Boxing   
3  2004  Summer          Athina   Wrestling   
4  2016  Summer  Rio de Janeiro    Swimming   

                                      Event Medal country  
0               Basketball Men's Basketball   NaN   China  
1              Judo Men's Extra-Lightweight   NaN   China  
2                 Boxing Men's Middleweight   NaN   China  
3  Wrestling Men's Lightweight, Greco-Roman   NaN   China  
4     Swimming Women's 200 metres Freestyle   NaN   China

**Question 3 :** Y a-t'il une ou des colonnes qui peut/peuvent servir d'index dans le DataFrame ? Autrement dit, quelle information utiliser pour identifier de manière unique chaque individu, au sens statistique du terme, de ce jeu de données ? Vous pouvez essayer de rechercher un de vos athlètes préférés pour vérifier cela.

In [4]:
# On pourrait indexer nos données à l'aide de la colonne 'ID'

raw_data.index = raw_data['ID']

In [5]:
# Vérifications pour répondre à la question :  chercher l'athlète Usain Bolt

usain_bolt = raw_data[raw_data['Name'].str.contains("Usain", na=False)]

usain_bolt[['ID','Name','Medal']]

ID                Name Medal
ID                                    
13029  13029  Usain St. Leo Bolt   NaN
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt   NaN
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold

=======> **Nous remarquons que l'ID unique de Usain Bolt est 13029, ce qui signifie qu'il est raisonnable d'indexer les athlètes par ID**

In [6]:
# Inversement, nous pouvons également chercher le nom de l'athlète correspondant à un ID donné

athlete23 = raw_data[raw_data.index == 23] #Athlète dont le ID est 23

athlete23.head()

ID         Name Sex   Age  Height  Weight    Team  NOC        Games  Year  \
ID                                                                              
23  23  Fritz Aanes   M  22.0   187.0    89.0  Norway  NOR  2000 Summer  2000   
23  23  Fritz Aanes   M  26.0   187.0    89.0  Norway  NOR  2004 Summer  2004   

    Season    City      Sport                                           Event  \
ID                                                                              
23  Summer  Sydney  Wrestling  Wrestling Men's Light-Heavyweight, Greco-Roman   
23  Summer  Athina  Wrestling  Wrestling Men's Light-Heavyweight, Greco-Roman   

   Medal country  
ID                
23   NaN  Norway  
23   NaN  Norway

=======> **L'athlète dont l'ID est 23 est le norvégien Fritz Aanes** 

## 2. Nettoyage des données

**Question 4 :** Y a-t-il des doublons dans ce jeu de données ? Si oui, proposez une solution pour supprimer les doublons.

In [7]:
# Vérification de la présence de doublons et proposition de solution

data = raw_data.drop_duplicates()

print(len(raw_data)) # nombre de toutes les données, les dupliquées incluses

print(len(data)) # nombre de données non dupliquées 

print(len(raw_data)-len(data)) # nombre de données dupliquées

283652
269382
14270


=======> **Il y avait donc 14270 données dupliquées qu'on a a présent supprimées pour avoir notre nouveau dataset data**

**Question 5 :** Les informations dont on dispose sont-elles complètes pour tous les athlètes ? Si des valeurs sont manquantes, pouvez-vous l'expliquer ?

On rappelle que d'un point de vue informatique, une case vide dans le csv est interprétée comme un `Nan` par la machine

In [8]:
# Vérification de l'absence de valeurs dans le jeu de données et proposition de solution

print(len(data))

DATA2 = data.dropna(how='any')

print(len(DATA2))

269382
30172


=======> **Quand nous enlevons toutes les lignes dans lesquelles il manque une des valeurs, il en reste 30172 sur 269382**

Cependant, enlever toutes ces valeurs n'est pas un choix rationnel. En effet, on remarque que pour la plupart des athlètes le champ **"Medal"** est vide car ils n'ont remporté aucune médaille. Ainsi, tout ce qu'il nous resterait dans notre base de données seraient les athlètes avec médaille. Or les athlètes qui n'ont pas remporté de médailles sont aussi importants.

=======> **Par conséquent, nous remplacerons dans la colonne "Medal" tous les NaN par des 0.** On pourra par la suite dans une future séance envisager, si nécessaire, d'attribuer des valeurs numériques pour le reste des médailles, par ex:  Bronze 1 Silver 2 et Gold 3

In [9]:
# remplacer toutes les NaN par des 0

data["Medal"].fillna(0, inplace = True)

print(len(data))

data.head()[["ID","Name","Medal"]]

# le paramètre "inplace = True" est important pour remplacer l'ancien dataset 'data' par sa nouvelle version

269382


/opt/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py:6245: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


ID           Name Medal
ID                             
1        1      A Dijiang     0
2        2       A Lamusi     0
602    602  Abudoureheman     0
1463  1463      Ai Linuer     0
1464  1464      Ai Yanhan     0

In [10]:
# Nous remarquons bien que les NaN ont été remplacées par des 0

usain_bolt = data[data['Name'].str.contains("Usain", na=False)]

usain_bolt[['ID','Name','Medal']]

# En prenant l'exemple de Usain Bolt on peut s'assurer que l'on a pas mis des 0 partout

ID                Name Medal
ID                                    
13029  13029  Usain St. Leo Bolt     0
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt     0
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold
13029  13029  Usain St. Leo Bolt  Gold

A présent, nous pouvons enlever toutes les lignes où il y a des valeurs manquantes car nous pouvons pas reproduire pour les autres colonnes ce qu'on a fait pour **"Medal"**. En effet, des données telles que **"Age"**, **"Weight"** et **"Height"** ne peuvent pas valoir 0 d'où la nécessité **d'enlever les lignes avec valeurs manquantes**.

In [11]:
# Enlever le reste des lignes avec valeurs manquantes (sans compter "Medal") 

DATA3 = data.dropna(how='any')

print(len(data)) # nombre de données avant

print(len(DATA3)) # nombre de données final

269382
205882


=======> **Nous avons donc une base de données "nettoyée" de 205882 lignes après avoir enlevé les doublons et les lignes avec valeurs manquantes (sans prendre en compte les médailles car cela nous aurait induits en erreur)**